File: Interview Cake/BalancedBinaryTree/BalancedBinaryTree.java
import java.util.*;
public static class BinaryTreeNode {
    public int value;
    public BinaryTreeNode left;
    public BinaryTreeNode right;

    public BinaryTreeNode(int value) {
        this.value = value;
    }

    public BinaryTreeNode insertLeft(int leftValue) {
        this.left = new BinaryTreeNode(leftValue);
        return this.left;
    }

    public BinaryTreeNode insertRight(int rightValue) {
        this.right = new BinaryTreeNode(rightValue);
        return this.right;
    }

}

class myCode {

    public ArrayList<Integer> getAllDepths(BinaryTreeNode root) {
        Stack holder = new Stack();
        BinaryTreeNode currentNode = root;
        ArrayList<Integer> depthHolder = new ArrayList<Integer>();
        while (true) {
            holder.push(currentNode);
            if (currentNode.left == null) {
                if (currentNode.right == null) {

                }
                holder.pop();

                currentNode = holder.top();
            }
            currentNode = currentNode.left;


        }
    }

	public static void main (String[] args) throws java.lang.Exception
    {
        //Current thought: Check super balanced by doing DFS

    }
}

File: Interview Cake/BalancedBinaryTree/balanced_binary_tree.py
  class BinaryTreeNode:

    def __init__(self, value):
        self.value = value
        self.left  = None
        self.right = None

    def insert_left(self, value):
        self.left = BinaryTreeNode(value)
        return self.left

    def insert_right(self, value):
        self.right = BinaryTreeNode(value)
        return self.right

def is_balanced(tree_root):
    depths = [] 
    nodes = [] 
    nodes.append((tree_root, 0))
    while len(nodes):
        node, depth = nodes.pop() 

        if (not node.left) and (not node.right) :
        #we found a leaf
            if depth not in depths:
                depths.append(depth)
                if (len(depths) > 2) or abs(depths[0] - depths[1]) > 1): 
                    return False

        else:
            if node.left:
                nodes.append(node.left)
            if node.right: 
                nodes.append(node.right)
    return True; 